# Task 4.7 Deriving New Variables

# Importing Libraries

In [9]:
# import libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [10]:
# importing orders_products_merged
orders_products_merged = pd.read_pickle(r'C:\Users\littl\Career Foundry\06.2022 Instacart Basket Analysis\02 - Data\Prepared Data\orders_products_merged.pkl')


In [11]:
# checking output
orders_products_merged.shape

(32406041, 16)

# Creating a Subset of Data

In [12]:
# creating a subset of the first 1,000,000 rows
df = orders_products_merged[:1000000]

In [13]:
df.shape

(1000000, 16)

# IF Statements w/ User Defined Function
Creating a function for the following price_labels within the subset of data:
* If prices are less than or equal to 5, then it is a 'Low-range product
* If prices are greater than 5, and less then or equal to 15, it is a 'mid range product'
* If prices are greater than 15,  it is a high-range product. 


In [14]:
# define a function
def price_label(row):
    
    if row['prices'] <= 5:
        return 'low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'mid range product'
    elif row['prices'] > 15:
        return 'high range product'
    else: return np.nan

In [15]:
# apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\littl\AppData\Local\Temp\ipykernel_408\1199080080.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [16]:
# check output
df['price_range'].value_counts(dropna = False)

mid range product    756450
low range product    243550
Name: price_range, dtype: int64

In [17]:
# checking max price as there is no foods that are listed as high-range product. 
df['prices'].max()

14.8

# IF Statements using the loc() Function
Using the location function to sort price ranges across the entire dataframe.

In [18]:
# orders greater than $15
orders_products_merged.loc[orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
# orders less than $5
orders_products_merged.loc[orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
# orders greater than 5 and less than 15
orders_products_merged.loc[(orders_products_merged['prices'] > 5) & (orders_products_merged['prices'] <= 15), 'price_range_loc'] = 'Mid-range product'

In [21]:
# checking output
orders_products_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21861997
Low-range product     10126366
High-range product      417678
Name: price_range_loc, dtype: int64

# IF Statements with For-Loops
Creating and defining a for loop function to display the busiest day, slowest day, and regular busy.

In [22]:
orders_products_merged.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'Unnamed: 0', 'product_name',
       'aisle_id', 'department_id', 'prices', '_merge', 'price_range_loc'],
      dtype='object')

In [23]:
# finding out the frequency for each day of the week, to see when the most orders take place.  
orders_products_merged['order_dow'].value_counts(dropna = False)

0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: order_dow, dtype: int64

In [24]:
# creating for loop for days of the week. 
result = []
for value in orders_products_merged["order_dow"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [26]:
orders_products_merged.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

In [33]:
# combining the for loop result with the dataset orders_prodcuts_merged
orders_products_merged['Busiest day'] = result

In [34]:
# counting values 
orders_products_merged['Busiest day'].value_counts(dropna = False)

Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: Busiest day, dtype: int64

In [37]:
# checking output
orders_products_merged.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     Busiest day  
0  Mid-range product  Regularly busy  
1  Mid-range product  Regularly busy  
2  Mid-range product  Regularly busy  
3  Mid-range product      Least busy  
4  Mid-range product      Least busy

# Task Questions

#### Question 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [46]:
# Create a new variable 'busiest days'
results2 =[]

for value in orders_products_merged['order_dow']:
    if value in [0,1]:
        results2.append("Busiest days")
    elif value in [3,4]:
        results2.append("Slowest days")
    else:
        results2.append("Regular days")

In [47]:
# combihe for loop request and column with count results
orders_products_merged['Busiest days'] = results2
orders_products_merged['Busiest days'].value_counts(dropna = False)

Regular days    12916527
Busiest days    11864860
Slowest days     7624654
Name: Busiest days, dtype: int64

In [49]:
orders_products_merged.shape

(32406041, 19)

We have successfully accounted for all the rows that are in the dataset.   12916527+11864860+7624654 = 32406041

#### Question 3: Check the values of this new column for accuracy. Note any observations in markdown format. 

Busiest days: Summing up the frequency of 0 and 1, match what is considered the total of busiest days (6204404 + 5660456 = 11,864,112)

Slowest Days: Summing up the frequency of 3 and 4, match what is considered the total for slowest days (3840701 + 3783953) = 7,624,654

Regular Days: Summing up the frequency of 6, 2, 5, match what is considered the total for Regular days (4496635 + 4213986 + 4205906) = 12,916,527

#### Question 4: When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

# determining the busiest hours of the day
df_busytime = orders_products_merged['order_hour_of_day'].value_counts()

In [53]:
# checking output
df_busytime

10    2761862
11    2736242
14    2689251
15    2662246
13    2661057
12    2618613
16    2535313
9     2454273
17    2087721
8     1718182
18    1636566
19    1258352
20     976182
7      891082
21     795658
22     634247
23     402335
6      290500
0      218786
1      115706
5       87964
2       69377
4       53245
3       51281
Name: order_hour_of_day, dtype: int64

In [60]:
# Create a new variable 'busiest period of the day'
results3 =[]

for value in orders_products_merged['order_hour_of_day']:
    if value >= 9 and value <=16:
        results3.append("Most Orders")
    elif value >=2 and value <=5:
        results3.append("Fewest Orders")
    else:
        results3.append("Average Orders")

In [62]:
# combininging the dataset with the forloop. 
orders_products_merged['Most Orders'] = results3

In [63]:
# checking output
orders_products_merged['Most Orders'].value_counts(dropna = False)

Most Orders       21118857
Average Orders    11025317
Fewest Orders       261867
Name: Most Orders, dtype: int64

In [64]:
orders_products_merged.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     Busiest day  Busiest days     Most Orders  
0  Mid-range product  Regularly busy  Regular days  Average Orders  
1  Mid-range product  Regularly busy  Slowest days  Average Orders  
2  Mid-range product  Regularly busy  Slowest days     Most Orders  
3  Mid-range product      Least busy  Slowest days  Average Orders  
4  Mid-range product      Least busy  Slowest days     Most Orders

Most orders are between 9am and 4pm; and the fewest orders are between 2am - 5am.

In [65]:
#Export data
orders_products_merged.to_pickle(r'C:\Users\littl\Career Foundry\06.2022 Instacart Basket Analysis\02 - Data\Prepared Data\4.7_orders_products_merged.pkl')
